src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs

```diff
@@ -20,7 +20,14 @@ public RemoveOperation(OperationBuilder builder, LazyItemEvaluator<P, I, M, D> l
             // todo Perf: do not match against the globs: https://github.com/Microsoft/msbuild/issues/2329
             protected override ICollection<I> SelectItems(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
             {
-                return SelectItemsMatchingItemSpec(listBuilder, _itemElement.RemoveLocation).ToImmutableHashSet();
+                var items = ImmutableHashSet.CreateBuilder<I>();
+                foreach (ItemData item in listBuilder)
+                {
+                    if (_itemSpec.MatchesItem(item.Item))
+                        items.Add(item.Item);
+                }
+
+                return items.ToImmutableHashSet();
```

Comment:
@cdmihai Any ideas why this is an immutable hashset?

Reply:
Because it's main use is items.Contains in SaveItems

---
In reply to: [143151572](https://github.com/Microsoft/msbuild/pull/2600#discussion_r143151572) [](ancestors = 143151572)